### DISCUSSION TITLE
Tasks/scripts without Projen?

### DISCUSSION THREAD

**Discussion by vaughandroid** [projen/projen#3159] [General]
As Node developers, my team and I are used to working with scripts that are defined in package.json. If we want to use Projen, all the (p)npm scripts just execute Projen tasks, and we are forced down the route of using and understanding Projen's task model (and tasks.json).

I can't see a compelling reason for this to be the case. I'd just like to use Projen for (re-)generating projects, not running my tasks. This makes adoption harder for teams, particularly since the "out of the box" task setups are quite rich and complicated.

1. I'm curious what the rationale is for putting Projen in charge of running tasks, rather than leaving it up to the various existing build systems that people will be familiar with.
2. Is there a way to get back to scripts without Projen in a Projen-generated project?
3. Is there a simple way to clear the standard set of tasks? (`project.tasks.all.forEach(task => project.removeTask(task.name))` errors because tasks need to be removed in a specific order.)

**Comment by mrgrain** [comment]
Thanks for your question and thoughts on this.

1. I think you are mixing two concepts into one here. Which is unsurprising, because in projen they are currently the same as well. The distinction is important, because each has a different answer.
Firstly, we have the concept of Components defining Tasks. That's a core functionality on probably not going away. Let's take the build and publish components as an example. Without tasks, these are much less useful. Now I can understand that someone might prefer (core) Components that don't do that, but that's not currently on the table.
Secondly, we have the concept of projen as a task runner. I don't know for sure (cause I wasn't around when this happened), but I imagine the reason for this was to come up with a simple task running solution that works the same across all languages. It's also nice to only have to remember to run `projen build` no matter if this is a npm, yarn or python project. But I guess opinions differ. Either way, it fits the "batteries included" mission. The good news for you is that we are well aware of the need to use different tasks runners, including npm. The bad news is that while it's planned, there's no current timeline for it. It [didn't quite make the cut for GA](https://github.com/projen/projen/issues/2976), but is high on the list for work afterwards.

2. Currently not really. You could use `projen eject` but then you still have to convert `tasks.json` to `package.json`.

3. I don't think that would be good idea. Many components depend on the tasks they register. What's the goal of removing all tasks? Are you mostly interested in cleaning up `package.json? If so, that might be possible. Probably using the low-level [ObjectFile API](https://projen.io/escape-hatches.html#object-file-patches). But I'll need to run some checks next week to confirm this.
